### ISSUE TITLE
Failure when using prune on a nested field

### ISSUE THREAD

**Issue by 3ygun** [logstash-plugins/logstash-filter-prune#12]
## Issue

Pruning nested fields within `whitelist_names` & `blacklist_names` doesn't work.

## Need?

I have a large amount of fields (>100) which are being processed. I want to prune down to a subset of those fields. However, the data looks as follows with a nested field having the same name as a non-nested field:

```
{
  ...
  "blam": "blam message",
  "bar": "bar message",
  ...
  "foo": {
     ...
    "blam": "foo blam message",
    "bar": "foo bar message",
     ...
  },
  ...
}
```
If I want to keep only `[foo][bar]` and `[blam]` how do I reference them while removing the rest?

## Test

I have the test: `{"hi":"hello","test":{"bar":"1","ble":2}}`

I want to run:

```
input {
  stdin {
    codec => json
  }
}

filter {
  prune {
    interpolate => true
    # Lets remove test.bar!
    blacklist_names => [ "[test][bar]" ]
  }
}

output {
  stdout { codec => rubydebug { metadata => true } }
}
```

But it fails to do anything:
![prune on nested field](https://cloud.githubusercontent.com/assets/5482320/17450279/2c5c584a-5b14-11e6-8182-08f22219c231.PNG)

I don't know if it's related to this issue: [Update remaining plugins with Event API](https://github.com/elastic/logstash/issues/5309)

**Comment by magnusbaeck** [comment]
Thanks for the clear report. You're probably right that it's the event API
that's the problem. I'm currently away and won't be able to look at this
until a day or two into next week though.

**Comment by 3ygun** [comment]
@magnusbaeck Thanks!

**Comment by magnusbaeck** [comment]
Adapting the plugin to the new event API turned out to be harder than I thought.

The `(white|black)list_names` options support field name regexps, but what does that even mean for nested fields (that so far have been unsupported by this plugin)? Should we generate a list of `field`, `[field]`, and `[field][subfield]` strings and match each item of `(white|black)list_names` against them? I feel there be many dragons here.

**Comment by boernd** [comment]
Why not just support white/blacklisting of top-level fields? Then it would work the same as before right?

**Comment by 3ygun** [comment]
@boernd while that may work it doesn't solve the problem that, from a user perspective, the behaviour is inconsistent with the rest of logstash. I wasted at least an hour on the above as I _expected_ it to work and it wasn't really documented otherwise.

@magnusbaeck notes below:
- As you have to add the `interpolate => true` flag to get the desired behaviour of `%{field}` behaviour such a list should only be computed when this flag is set. Furthermore, the `interpolate` documentations mentions it adds performance problems.
- However, after re-looking into the problem there might be deeper issues with getting even `[field]` to work. The `[  ]` in regex, I believe is a "character group", thus determining whether the user wants a "character group" or a logstash field name could become near impossible.
- There needs to be better consistency within the prune plugin or at the very least better documentation. However, the primary reason I used the plugin was for the `(white|black)list_names` which doesn't have replacement functionality in another filter plugin.
- I've also added clarification about the use-case above

**Comment by boernd** [comment]
I just realized this filter does not do anything in LS version 5.0.0 (not even filtering "normal"/non-nested fields. Adding this line seems to fix it but I'm not sure how this could work before (old api):

```
...
    event.overwrite(LogStash::Event.new(hash))

    filter_matched(event)
...
```

**Comment by AlexClineBB** [comment]
I also have a similar issue. I have an unknown number of subfields that are named `*_url` which I'd like to discard. I'd love for the following prune configuration to remove them from the event:
```
prune {
  blacklist_names => [ "^.*_url$" ]
}
```

**Comment by 3ygun** [comment]
@magnusbaeck Although #19 does look like it solves the documentation problem. The docs/guide that I was consulted throughout my use of logash were the ones on elastic.co which don't appear to have/show the note: [prune filter](https://www.elastic.co/guide/en/logstash/current/plugins-filters-prune.html). Just wanted to point the difference out. e.g. I didn't look at the source code of the filters while using them (probably bad, but what I did).

**Comment by magnusbaeck** [comment]
> Although #19 does look like it solves the documentation problem. The docs/guide that I was consulted throughout my use of logash were the ones on elastic.co which don't appear to have/show the note

Yes, unfortunately the docs published on elastic.co are often behind. The Elastic folks are working on a rewrite of how they're generated.

**Comment by 3ygun** [comment]
Thanks for the information. Glad to hear.

**Comment by Popsiclestick** [comment]
This ticket has been opened a long time. Is there a plan to address this issue in a feature request somewhere? or will pruning in nested fields never be supported?

**Comment by magnusbaeck** [comment]
At this point it's not clear what a solution would look like, i.e. what semantics make sense for nested fields. Once there's a clear path an implementation is probably not that difficult.

Would it be reasonable to have an optional configuration option named e.g. `base_field` (default empty/null) that chooses which field or subfield to use as the base for the field or value blacklist? That way you'd do

```plain
prune {
  whitelist_names => ["foo", "bar"]
  base_field => "toplevel_field"
}
```

to only keep `[toplevel_field][foo]` and `[toplevel_field][bar]` of the total set of subfields of `toplevel_field`. All other fields at the top level (and their subfields) would be left alone. Now, this would prevent you from whitelisting or blacklisting all fields with a particular name regardless of where they're found in the field hierarchy but the feature would be easy to understand and fully backwards compatible.

**Comment by fozboz** [comment]
Could `base_field` also be a nested value?
```
prune {
  whitelist_names => ["foo", "bar"]
  base_field => "[toplevel_field][nextlevel_field][yet_another_level]"
}
```
For now, I'm working round this with
```
mutate {
  copy => { "[the][only][thing][i][want]" => "what_i_want" }
}
prune {
  whitelist_names => [ "what_i_want" ]
}
```
but that's not helpful if you need to do a regex.

**Comment by magnusbaeck** [comment]
> Could `base_field` also be a nested value?

Yes, when `base_field` is implemented I see no reasons why it can't address a subfield.

**Comment by reuschling** [comment]
I have the problem that I want to whitelist some attributes, but then the metrics plugin stopped working because the attributes in the events generated by metrics are also nested - and will be pruned away gracefully.  Sadly, an if-statement around prune that it only prunes away the events from the input doesn't work which is strange to me, 'type' is set to 'input' in the event:
``
if [type] == "input"{
prune { whitelist_names =>["^tags$", "^title$", "^rate_15m$", "^rate_1m$", "^count$", "^rate_5m$", "^@timestamp$", "^@version$", "^message$","events"]
		}
    }
``

**Comment by Mike-Logit** [comment]
> Yes, when `base_field` is implemented 

This request was created in 2016. I doubt it's going to be implemented anytime soon, if ever. Sad because not having the ability to use nested fields significantly limits this plugin's capabilities.

**Comment by magnusbaeck** [comment]
> I have the problem that I want to whitelist some attributes, but then the metrics plugin stopped working because the attributes in the events generated by metrics are also nested - and will be pruned away gracefully.

This probably isn't a problem with the plugin and is only tangentially related to this issue, so please open a thread at https://discuss.elastic.co/c/logstash/. I don't track new threads there anymore but post the URL here and I'll have a look.

**Comment by Mike-Logit** [comment]
The issue with base_field (in terms of semantics) is whether it will remove all other fields (including nested fields) in the entire json document or whether it just means remove all fields in the current nested field targeted by base_field.

I thought it would be more useful to simply specify the path of the field to keep, relative to the entire document, as an item in the array:

```
prune {
    whitelist_names => [ 
        "top_field.another_field.test_(a|b|c)$", # some regex here
        "top_field.different_field.something", # a different 2nd nested field
        "keepme", # top level
    ]
}
```

**Comment by 3ygun** [comment]
> > Yes, when `base_field` is implemented
>
> This request was created in 2016. I doubt it's going to be implemented anytime soon, if ever. Sad because not having the ability to use nested fields significantly limits this plugin's capabilities.

When I opened the ticket is was because of the limiting nature. I don't exactly remember how I got around this limitation? Probably, realized I didn't need the functionality because I hit into https://github.com/elastic/kibana/issues/1084 or did something like fozboz's https://github.com/logstash-plugins/logstash-filter-prune/issues/12#issuecomment-434454754. My entire project at the time was visualizing the data. Thus if it cannot be visualized in a nested format why keep it in the nested format at all? As such, to avoid the Kibana side limitation I probably just flattened/remapped the log messages to a flat object and then pruned from there.

Overall, @magnusbaeck:
- My original use case has passed (no longer with the company)
- The limitation on the Kibana side would be more pressing to that original use case anyway

I think the `base_field` idea would be a good middle ground (between not having and an implementation on `blacklist_names` which shouldn't happen for backwards compatibility?). If you could possibly through a regex in `base_field` I think that would solve everything.

Believe the following would be a possible example (sorry been out of the flow for a while):

Input
```json
{
    "message_header": {
        "name": "Document Upload",
        "type": "POST",
        "object": ".txt"
    },
    "message_body": {
        "name": "hello.txt",
        "type": "document",
        "object": "...base64 encoded file....",
        "from": "David Soller"
    },
    "message_auth": {
        "name": "OAuth",
        "type": "client_credentials",
        "object": "client_id=123&client_secret=secret",
        "from": "Patient 123 - David Soller"
    }
}
```

Filter
```ruby
prune {
  blacklist_names => [ "object", "from" ]
  base_field => "message_(auth|body)"
}

prune {
  blacklist_names => [ "name" ]
  base_field => "message_body"
}
```

To get final message
```ruby
{
    "message_header": {
        "name": "Document Upload",
        "type": "POST",
        "object": ".txt"
    },
    "message_body": {
        "type": "document"
    },
    "message_auth": {
        "name": "OAuth",
        "type": "client_credentials"
    }
}
```

Basically you're trying to prune the possible patient identifiable information from the original message and just keep the "type" details about how people are using this "Document Upload". Apologies if it is too contrived and/or already handled in Logstash.

**Comment by pemontto** [comment]
So 3 years later this still isn't working, or am I just doing something wrong? Just tested on v 7.3.2 😞 

```
filter {
  prune {
    whitelist_names => [ "delivered-to", "message", "from", "to", "date"]
  }
}
```

**Comment by developedsoftware** [comment]
Was hoping to use prune to remove many null values in nested objects rather than loop through with a script. 

Could we not simply add extra paramters to prune called path_match and path_unmatch like we do in dynamic mapping?  

So you could do *.name to prune ever name field inside each top level object or *.* (to prune everything inside every top level object)

**Comment by immon** [comment]
As a workaround I came up with a ruby script to whitelist nested fields.

```
filter {
    ruby {
        init => "
        @whitelist = ['@timestamp', 'server.ip','server.port','event.category']

        ####

        @whitelist_keys = Hash[@whitelist.collect { |item| [item, true] } ]
        @whitelist.each do |value|
            parts = value.split('.')
            prefix = ''
            parts.slice(0..-2).each do |part|
                part = prefix + '.' + part unless prefix.empty?
                @whitelist_keys[part] = false
                prefix = part
            end
        end

        # puts @whitelist_keys
        "

        code => "

            def reduce(hash, prefix='')
                prefix = prefix + '.'
                new_hash = {}
                hash.each do |key, value|
                    full_key = prefix + key
                    if value.is_a?(Hash)
                        new_hash[key] = reduce(value, full_key) if @whitelist_keys.has_key?(full_key)
                    else
                        new_hash[key] = value if @whitelist_keys[full_key]
                    end
                end
                # puts new_hash
                new_hash
            end

            event.to_hash.each do |key, value|
                if value.is_a?(Hash)
                    event.set(key, reduce(value, key)) if @whitelist_keys.has_key?(key)
                else
                    event.remove(key) unless @whitelist_keys[key]
                end
            end

        "
    }
} 
```

**Comment by fabrei** [comment]
Would be a helpful feature! At the moment I have a workaround for blacklisting:

```
rename => {
    "[something][for][garbage]" => "garbage_something"
}
prune {
    blacklist_names => ["garbage_something"]
}
```
But it is a bit nasty if I have a lot of fields to blacklist; the config gets bigger and bigger..

**Comment by hackery** [comment]
Same issue here. We have a pipeline for web logs which splits a query string using [kv](https://www.elastic.co/guide/en/logstash/current/plugins-filters-kv.html) into individual `param.foo` fields, but that can be an arbitrarily long list (especially if some automated vulnerability scan starts throwing garbage at your web service!). Sadly kv's `include_keys` doesn't accept a regex.

These params go into a subfield `params` obviously to avoid them mashing over any standard fields. The `base_field` option would be perfect here even in its simplest form.

**Comment by SudoNova** [comment]
Hi. I've got same issue. Pruning nested fields with `blacklist_names` doesn't work, I've tried `^log\.`, `^log.`, `^log.*` but no results.

**Comment by benmatthews** [comment]
@immon this ruby filter works for us but unknown fields containing hashes were not removed - suggest improvement

```
event.to_hash.each do |key, value|
  if value.is_a?(Hash)
      event.remove(key)
      if @whitelist_keys.has_key?(key)
          filtered = reduce(value, key)
          if !filtered.empty?
              event.set(key, filtered)
          end
      end
  else
      event.remove(key) unless @whitelist_keys[key]
  end
end 
```
